protocol.md

```diff
@@ -773,6 +773,47 @@ The `Upload-Tag` header MUST be included with the `POST` request that creates
 the resource. It MUST also be sent with a follow-up `HEAD` request used to retrieve the
 `Location` of the newly created resource.
 
+##### Upload-Tag-Secret
+
+The `Upload-Tag-Secret` MUST be a high-entropy cryptographic random [base64 alphabet](https://tools.ietf.org/html/rfc3986#section-2.3) string
+with a minimum length of 48 characters and a maximum length of 256 characters.
+
+For unauthenticated uploads, Clients MUST send the `Upload-Tag-Secret` header
+alongside the `Upload-Tag` header with the `POST` request that creates the upload
+resource.
+
+Servers MUST return a `400 Bad Request` status for unauthenticated `POST` requests
+that contain an `Upload-Tag` header, but do not also include a valid `Upload-Tag-Secret`
+header.
+
+For authenticated uploads, Servers MUST respond to `PATCH` and `HEAD` requests with
+a `404 Not Found` or `401 Unauthorized` status, if they include an `Upload-Tag` header
+identifying an upload resource that was created by a different user.
+
+Clients MAY include an `Upload-Tag-Secret` header for authenticated uploads, but are
+not REQUIRED to do so.
+
+##### Upload-Tag-Challenge
+
+If an `Upload-Tag-Secret` header was sent with the `POST` request that created
+the upload resource, Clients MUST include an `Upload-Tag-Challenge` header with every
+subsequent `HEAD` and `PATCH` request that also contains an `Upload-Tag`.
+
+The value of `Upload-Tag-Secret` is defined as follows:
+
+```
+Upload-Tag-Challenge = SHA256([Upload-Tag-Secret] + SHA256([Upload-Tag-Secret] + [Upload-Offset] + [Content-Length]))
```

Comment:
> I think using an HMAC with `Upload-Tag-Secret` as the key would make more sense.

I also considered HMACs, but eventually opted for SHA256 as it should be more widely available in various programming languages.

Seeing as f.ex. [OAuth2 PKCE also uses just SHA256](https://tools.ietf.org/html/rfc7636#section-4.1), I'd think it's safe for the intended purpose.

> Concatenating offset and length is ambiguous, we should have a separator character.

Missed that. Thanks for pointing that out! I shuffled the parameters around so that now the secret it between the two and acts as a separator.